File: day19.c
#include <stdio.h>
#include <string.h>

char pats[512][8 + 1];
int n;
long long memo[60 + 1];

long long solve(const char* target, int i)
{
    if (i == 0) memset(memo, -1, sizeof(memo));
    if (target[i] == 0) return 1;
    if (memo[i] >= 0) return memo[i];

    long long count = 0;
    for (int p = 0; p < n; p++) {
        int len = strlen(pats[p]);
        if (!strncmp(&target[i], pats[p], len)) {
            count += solve(target, i + len);
        }
    }

    return memo[i] = count;
}

int main()
{
    FILE* file = fopen("day19.txt", "r");
    long long sum = 0, sum2 = 0;

    while (1 == fscanf(file, "%[a-z],%*[ ]", pats[n])) n++;

    char target[60 + 1];
    while (1 == fscanf(file, "%s", target)) {
        long long count = solve(target, 0);
        sum += count != 0;
        sum2 += count;
    }

    printf("%lld %lld\n", sum, sum2);
}

File: day19.lua
#!/usr/bin/env lua

local pats = {}
local memo = {}

local function solve(target, i)
    if i == 1 then memo = {} end
    if i > #target then return 1 end
    if memo[i] then return memo[i] end

    local count = 0
    for _, pat in ipairs(pats) do
        local len = #pat
        if target:sub(i, i + len - 1) == pat then
            count = count + solve(target, i + len)
        end
    end
    memo[i] = count
    return count
end

local file = io.open("day19.txt")
local sum = { 0, 0 }

for x in file:read():gmatch("%l+") do
    pats[#pats + 1] = x
end

for target in file:read("a"):gmatch("%l+") do
    local count = solve(target, 1)
    if count ~= 0 then sum[1] = sum[1] + 1 end
    sum[2] = sum[2] + count
end

print(sum[1], sum[2])
